consensus/progpow/consensus.go

```diff
@@ -0,0 +1,526 @@
+					errorsOut <- errors[out]
+					if out == len(headers)-1 {
+						return
+					}
+				}
+			case <-abort:
+				return
+			}
+		}
+	}()
+	return abort, errorsOut
+}
+
+func (progpow *Progpow) verifyHeaderWorker(chain consensus.ChainHeaderReader, headers []*types.Header, seals []bool, index int, unixNow int64) error {
+	var parent *types.Header
+	if index == 0 {
+		parent = chain.GetHeader(headers[0].ParentHash(), headers[0].NumberU64()-1)
+	} else if headers[index-1].Hash() == headers[index].ParentHash() {
+		parent = headers[index-1]
+	}
+	if parent == nil {
+		return consensus.ErrUnknownAncestor
+	}
+	return progpow.verifyHeader(chain, headers[index], parent, false, seals[index], unixNow)
+}
+
+// VerifyUncles verifies that the given block's uncles conform to the consensus
+// rules of the stock Quai progpow engine.
+func (progpow *Progpow) VerifyUncles(chain consensus.ChainReader, block *types.Block) error {
+	// If we're running a full engine faking, accept any input as valid
+	if progpow.config.PowMode == ModeFullFake {
+		return nil
+	}
+	// Verify that there are at most 2 uncles included in this block
+	if len(block.Uncles()) > maxUncles {
+		return errTooManyUncles
+	}
+	if len(block.Uncles()) == 0 {
+		return nil
+	}
+	// Gather the set of past uncles and ancestors
+	uncles, ancestors := mapset.NewSet(), make(map[common.Hash]*types.Header)
+
+	number, parent := block.NumberU64()-1, block.ParentHash()
+	for i := 0; i < 7; i++ {
+		ancestorHeader := chain.GetHeader(parent, number)
+		if ancestorHeader == nil {
+			break
+		}
+		ancestors[parent] = ancestorHeader
+		// If the ancestor doesn't have any uncles, we don't have to iterate them
+		if ancestorHeader.UncleHash() != types.EmptyUncleHash {
+			// Need to add those uncles to the banned list too
+			ancestor := chain.GetBlock(parent, number)
+			if ancestor == nil {
+				break
+			}
+			for _, uncle := range ancestor.Uncles() {
+				uncles.Add(uncle.Hash())
+			}
+		}
+		parent, number = ancestorHeader.ParentHash(), number-1
+	}
+	ancestors[block.Hash()] = block.Header()
+	uncles.Add(block.Hash())
+
+	// Verify each of the uncles that it's recent, but not an ancestor
+	for _, uncle := range block.Uncles() {
+		// Make sure every uncle is rewarded only once
+		hash := uncle.Hash()
+		if uncles.Contains(hash) {
+			return errDuplicateUncle
+		}
+		uncles.Add(hash)
+
+		// Make sure the uncle has a valid ancestry
+		if ancestors[hash] != nil {
+			return errUncleIsAncestor
+		}
+		if ancestors[uncle.ParentHash()] == nil || uncle.ParentHash() == block.ParentHash() {
+			return errDanglingUncle
+		}
+		if err := progpow.verifyHeader(chain, uncle, ancestors[uncle.ParentHash()], true, true, time.Now().Unix()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// verifyHeader checks whether a header conforms to the consensus rules
+func (progpow *Progpow) verifyHeader(chain consensus.ChainHeaderReader, header, parent *types.Header, uncle bool, seal bool, unixNow int64) error {
```

Comment:
is it wise to have different implementation of these functions?, if its possible I would recommend having only the parts that are affected by a consensus engine separate. Looking at it most of the function is same for both progpow and blake3.

Reply:
> is it wise to have different implementation of these functions?

Isn't that the entire point of an interface? To define a set of methods that multiple modules may implement?

Yes, I think it makes sense to have multiple implementations. This makes it so we can switch consensus engines out or even support multiple parallel engines, e.g. if we decide to do a scheduled transition from ProgPow to Blake3pow.

^ to be clear, I'm not suggesting we should or will do such a transition, but we should keep the option open until we decide we don't need it.

Reply:
After the in person discussion, I agree with the current implementation now, since we are unsure about the direction yet